### ISSUE TITLE
Reaching 3.39 BPD on CIFAR

### ISSUE THREAD

**Issue by robintibor** [y0ast/Glow-PyTorch#5]
How to reach the 3.39 BPD on CIFAR10 written in the README using this code for training?
I thought calling `python train.py --download` should already be enough? I have started the training like that , seed used was 9376. After 103 epochs the validation nll printed is 4.01:
`Validation Results - Epoch: 103 total_loss: 4.03, nll: 4.01` And it does not seem to be substantially decreasing anymore (first reached nll 4.01 at epoch 86).

Do I have some misunderstanding, like is some other setting needed? the printed nll should be in bits per dimension right?

**Comment by robintibor** [comment]
Ah and along the same lines, could you supply bpd values for earlier epochs to allow for faster comparisons? so one can see whether training seems to be working as expected more quickly.

**Comment by robintibor** [comment]
Small update: seems possibly related to the warmup. When I remove the warmup, I also get NaNs for some seeds. But if there are no NaNs, after 4 epochs I already have BPD ~3.8, which is lower than what I got with warmup which is lower than the ~3.9 BPD I get with warmup  at end of training after 250 epochs. Will see where the runs without warmup will end tomorrow...

**Comment by robintibor** [comment]
Yes, it seems confirmed to me: now after ~55 epochs the runs without warmup have reached BPD ~3.5, so they may quite possibly end up around ~3.4 BPD. 
I noticed that without warmup, there are large increases in the loss in the first batch updates, and then fairly rapid decreases (according to the terminal outputs). Maybe this "unstable" phase is necessary to reach the lower BPDs later?

**Comment by y0ast** [comment]
Thanks for the updates @robintibor. It seems that indeed the warmup doesn't behave as hoped.

I'm looking into it.

**Comment by rmehta1987** [comment]
Could this be a problem, https://github.com/y0ast/Glow-PyTorch/blob/master/model.py#L78

`z2 = z2 + shift
z2 = z2 * scale`

should be:

`z2 = z2 * scale
z2 = s2 + shift`

**Comment by y0ast** [comment]
That isn't the problem. I follow the original glow implementation: https://github.com/openai/glow/blob/d3906b2f61183e2178feeb1d63dd5e5c9b78d04e/model.py#L377

It was a scaling issue with the learning rate. I'm nearly done rerunning the model training and will push the fix.

**Comment by rmehta1987** [comment]
That is strange as the coupling is different from the original paper and the Tensorflow implementation:

https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/bijectors/affine.py#L505

**Comment by robintibor** [comment]
I saw you made this fix: https://github.com/y0ast/Glow-PyTorch/commit/17848ce831b67c3d89200e525e0ace1c0b1630f7 did you run it to 250 epochs? Was this then reaching the 3.39 as well?  I always ran without warmup and reached close to 3.39...

**Comment by y0ast** [comment]
Thanks for the reminder @robintibor. I am playing around with settings so the result is both reliable and good.

Using warmup seems to hurt performance, while no warmup leads to NaNs sometimes. I'm running a few models now and will let you know.

**Comment by y0ast** [comment]
I've distilled the problem down to the optimizer. It turns out that `adamax` was vital, and that `adamax` works well with warmup too (`adam` is worse in general, but especially with warmup).

I have updated the code and added a comment to the readme. I do not quite know why `adamax` makes such a difference, so if anyone has an idea I'd be happy to hear it.

I'm closing this issue now as the code part is resolved :)
